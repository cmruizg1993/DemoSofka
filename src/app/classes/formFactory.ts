
import { IForm } from "../interfaces/i-form";
import { IFormField } from "../interfaces/i-form-field";

export abstract class FormFactory implements IForm{
  
  formFields: IFormField[] = [];

  getFormField(name: string): IFormField {
    let field = this.formFields.find( f => f.name == name );
    
    if(!field)
      field = {
        name
      };
    return field;
  }
  getCreationFormFields(): IFormField[] {
    throw new Error("Method not implemented.");
  }
  getUpdateFormFields(): IFormField[] {
    throw new Error("Method not implemented.");
  }
    
}
