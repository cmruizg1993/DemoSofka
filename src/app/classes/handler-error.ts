import { Observable, throwError } from "rxjs";
import { IError, IHandlerError } from "../interfaces/i-handler-error";
import { HttpErrorResponse } from "@angular/common/http";

export class HandlerError implements IHandlerError{
    
    handlerError(errorResponse: HttpErrorResponse): Observable<never> {
        console.log('OcurriÃ³ un error: ', errorResponse);
        return throwError(()=>{
            const error: IError ={
                name: errorResponse.name,
                message: errorResponse.message
            }
            if(!!errorResponse.error?.name)
                error.name = errorResponse.error.name;
            if(!!errorResponse.error?.message)
                error.message = errorResponse.error.message;
            if(!!errorResponse.error?.stack)
                error.stack = errorResponse.error.stack;
            return error;
        })
    }

}