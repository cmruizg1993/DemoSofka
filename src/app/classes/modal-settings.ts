import { Observable, Subject } from "rxjs";

export class ModalSettings {
    title: string = '';
    content: string = '';
    confirmButton = false;
    cancelButton = false;
    confirmButtonLabel: string = '';
    cancelButtonLabel: string = '';
    
    confirmAction?: CallableFunction;
    cancelAction?: CallableFunction; 

    showModal$ = new Subject<boolean>();
    get show$(): Observable<boolean>{
        return this.showModal$.asObservable();
    }
    open(){
        this.showModal$.next(true)
    }
    close(){
        this.showModal$.next(false);
    }
}
