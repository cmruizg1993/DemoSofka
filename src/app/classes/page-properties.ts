export class PagerProperties{
    page: number = 1;
    pageSize = 5;
  }