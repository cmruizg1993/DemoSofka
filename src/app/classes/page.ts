export class Page{
    number: number = 0;
    isActive: boolean = false;
  }