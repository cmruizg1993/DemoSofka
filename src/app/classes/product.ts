
import { IProduct } from "../interfaces/iproduct";

export class Product implements IProduct{

    id: string;
    name: string;
    description: string;
    logo: string;
    date_release: string;
    date_revision: string;

    constructor(
        id = '',
        name = '',
        description = '',
        logo: '',
        date_release: '',
        date_revision: ''
    ) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.logo = logo;
        this.date_release = date_release;
        this.date_revision = date_revision;
    }

}
