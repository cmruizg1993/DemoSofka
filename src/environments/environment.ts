export const environment = {
    apiUrl: '/bp'
};
