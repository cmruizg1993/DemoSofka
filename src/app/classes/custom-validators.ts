import { AbstractControl, AsyncValidatorFn, ValidationErrors, ValidatorFn } from "@angular/forms";
import { ProductService } from "../services/product.service";
import { Observable, map } from "rxjs";

export  class CustomValidators {
    static productIDValidator(productService: ProductService): AsyncValidatorFn {
        return (control: AbstractControl): Observable<ValidationErrors | null> => {
            const id = (control.value as string).trim().toLowerCase();
            const response$ = productService.verificationID(id);
            return response$.pipe(map(isExisting => (isExisting ? { idExists: true } : null)));
        };
    }
    static productValidateDateFormat(): ValidatorFn{
        return (control: AbstractControl): ValidationErrors | null => {
            const regex = /[0-9]{4,4}-[0-9]{2,2}-[0-9]{2,2}/;
            const date = new Date(control.value);
            return !regex.test(control.value) || isNaN(date.getDate()) ? { invalidDateFormat: { value: control.value } } : null;
        };
    }
    static productValidateDateValue(): ValidatorFn{
        
        return (control: AbstractControl): ValidationErrors | null => {
            const date = new Date(control.value);
            const currentDate = new Date();
            return date < currentDate ? { invalidDateValue: { value: control.value } } : null;
        };
    }
}
