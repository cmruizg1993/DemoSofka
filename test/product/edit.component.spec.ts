import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { CreateComponent } from '../../src/app/product/create/create.component';
import { ProductService } from '../../src/app/services/product.service';
import { HttpClientModule } from '@angular/common/http';
import { ComponentsModule } from '../../src/app/components/components.module';
import { FormComponent } from '../../src/app/components/form/form.component';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { environment } from '../../src/environments/environment.development';
import { EditComponent } from '../../src/app/product/edit/edit.component';
import { Router } from '@angular/router';
import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { Product } from '../../src/app/classes/product';
import { ButtonComponent } from '../../src/app/components/button/button.component';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { ModalComponent } from '../../src/app/components/modal/modal.component';




describe('Product/EditComponent', () => {
  let service: ProductService;
  let fixture: ComponentFixture<EditComponent>;
  let compiled: HTMLElement;
  let httpMock: HttpTestingController;
  let router: Router;

  const dateNow = new Date();
    const dateAfterYear = new Date();
    dateAfterYear.setFullYear(dateAfterYear.getFullYear() + 1); 
    const id = `${Math.floor(Math.random()*999999999)}`;

    const product: any = {
      id: id,
      name: `Test product ${id}`,
      description: 'DescripciÃ³n ...',
      logo: 'assets-1.png',
      date_release: `${dateNow.toISOString().split('T')[0]}`,
      date_revision: `${dateAfterYear.toISOString().split('T')[0]}`
    }

  beforeEach(async () => {
    

    await TestBed.configureTestingModule({
      imports: [
        RouterTestingModule.withRoutes([
          {
            path: 'edit',
            component: EditComponent, 
            data: product
          }
        ]),
        HttpClientTestingModule, FormsModule, ReactiveFormsModule
      ],
      declarations: [
        EditComponent, FormComponent, ButtonComponent, ModalComponent
      ],
      providers:[
        ProductService,
      ]
    }).compileComponents();
    
    router = TestBed.inject(Router);

    jest.spyOn(router, "getCurrentNavigation").mockReturnValueOnce({ extras: { state: product } } as any);

    service = TestBed.inject(ProductService);
    service.urlBase = environment.apiUrl;
    httpMock = TestBed.inject(HttpTestingController);
    fixture = TestBed.createComponent(EditComponent);
    
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('Se debe crear el Product/EditComponent', () => {
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });

  it(`Debe coincidir con el snapshot'`, () => {
    expect(compiled).toMatchSnapshot();
  });
  it(`Debe renderizar el form Component'`, () => {
    const component = fixture.componentInstance;
    fixture.detectChanges();
    const form = compiled.querySelector('app-form>form');
    const inputs = form?.querySelectorAll('input');
    expect(inputs?.length).toBe(6);
  });
  it(`El formulaio debe contener el ID : ${product.id}'`, () => {
    const component = fixture.componentInstance;
    fixture.detectChanges();
    
    const form = compiled.querySelector('app-form>form');
    const inputs = form?.querySelectorAll('input');
    expect(inputs?.item(0).value).toBe(product.id);
  });
  it(`Debe editar un producto'`, () => {
    
    const component = fixture.componentInstance;
    
    const formElement: DebugElement = fixture.debugElement.query(By.directive(FormComponent));

    jest.spyOn(component, 'updateProduct');    

    formElement.componentInstance.onClick(component.buttons[0]);

    fixture.detectChanges();

    expect(component.updateProduct).toHaveBeenCalledWith(product);

    const request = httpMock.expectOne(`${environment.apiUrl}/products/${product.id}`);
    expect( request.request.method ).toBe('PUT');
    request.flush( product );
    
  });
  it(`Debe fallar al editar un producto'`, (done) => {

    const product: any = {
      id: '1'
    }

    const component = fixture.componentInstance;

    component.updateProduct(product).then(()=>{
      let errResponse = component.modalSettings.content;
      expect(errResponse).toMatch('Hubo un error al actualizar el producto.');
      done();
    });
    const mockErrorResponse = { status: 400, statusText: 'Bad Request' };
    const data = 'Invalid request parameters';
    const request = httpMock.expectOne(`${environment.apiUrl}/products/${product.id}`);
    expect( request.request.method ).toBe('PUT');
    request.flush(data, mockErrorResponse);
    fixture.detectChanges();
    
  });

});

describe('Product/EditComponent', () => {
  let service: ProductService;
  let fixture: ComponentFixture<EditComponent>;
  let compiled: HTMLElement;
  let httpMock: HttpTestingController;
  let router: Router;

  beforeEach(async () => {
    

    await TestBed.configureTestingModule({
      imports: [
        RouterTestingModule.withRoutes([
          {
            path: 'edit',
            component: EditComponent
          }
        ]),
        HttpClientTestingModule, FormsModule, ReactiveFormsModule
      ],
      declarations: [
        EditComponent, FormComponent, ButtonComponent, ModalComponent
      ],
      providers:[
        ProductService,
      ]
    }).compileComponents();
    
    router = TestBed.inject(Router);

    //jest.spyOn(router, "getCurrentNavigation").mockReturnValueOnce({ extras: { state: product } } as any);    
  });

  it('Fields deben ser undefined', () => {
    let error = null;
    try{
      service = TestBed.inject(ProductService);
      service.urlBase = environment.apiUrl;
      httpMock = TestBed.inject(HttpTestingController);
      fixture = TestBed.createComponent(EditComponent);      
      fixture.detectChanges();
    }catch(err){
      error = err
    }
    expect(error).toBeTruthy();
  });


});

export function findComponent<T>(
  fixture: ComponentFixture<T>,
  selector: string,
): DebugElement {
  return fixture.debugElement.query(By.css(selector));
}
