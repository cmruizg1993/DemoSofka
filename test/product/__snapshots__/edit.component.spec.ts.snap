// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product/EditComponent Debe coincidir con el snapshot' 1`] = `
<div
  id="root1"
>
  <div
    class="row mt-2 justify-content-center"
  >
    <div
      class="col-md-6 bg-white"
    >
      <div
        class="row p-2 justify-content-center"
      >
        <p
          class="text-primary fs-4 fw-800"
        >
          Formulario de Edici贸n
        </p>
        <div
          class="hr mt-2"
        />
      </div>
      <app-form>
        <form
          class="ng-untouched ng-pristine ng-invalid"
          novalidate=""
        >
          <div
            class="row"
          >
            <div
              class="col-md-6"
            >
              <div
                class="form-group p-2"
              >
                <label
                  class="text-primary fs-2 fw-800"
                >
                  ID
                </label>
                <input
                  class="w100 ng-untouched ng-pristine"
                  disabled=""
                  placeholder=""
                  type="text"
                />
                <div
                  class="text-danger fw-bolder"
                  style="height: 1rem !important;"
                >
                  
                </div>
              </div>
            </div>
            <div
              class="col-md-6"
            >
              <div
                class="form-group p-2"
              >
                <label
                  class="text-primary fs-2 fw-800"
                >
                  Nombre
                </label>
                <input
                  class="w100 ng-untouched ng-pristine ng-valid"
                  placeholder=""
                  type="text"
                />
                <div
                  class="text-danger fw-bolder"
                  style="height: 1rem !important;"
                >
                  
                </div>
              </div>
            </div>
            <div
              class="col-md-6"
            >
              <div
                class="form-group p-2"
              >
                <label
                  class="text-primary fs-2 fw-800"
                >
                  Descripci贸n
                </label>
                <input
                  class="w100 ng-untouched ng-pristine ng-invalid"
                  placeholder=""
                  type="text"
                />
                <div
                  class="text-danger fw-bolder"
                  style="height: 1rem !important;"
                >
                  
                </div>
              </div>
            </div>
            <div
              class="col-md-6"
            >
              <div
                class="form-group p-2"
              >
                <label
                  class="text-primary fs-2 fw-800"
                >
                  Logo
                </label>
                <input
                  class="w100 ng-untouched ng-pristine ng-valid"
                  placeholder=""
                  type="text"
                />
                <div
                  class="text-danger fw-bolder"
                  style="height: 1rem !important;"
                >
                  
                </div>
              </div>
            </div>
            <div
              class="col-md-6"
            >
              <div
                class="form-group p-2"
              >
                <label
                  class="text-primary fs-2 fw-800"
                >
                  Fecha liberaci贸n
                </label>
                <input
                  class="w100 ng-untouched ng-pristine ng-invalid"
                  placeholder="yyyy-mm-dd"
                  type="text"
                />
                <div
                  class="text-danger fw-bolder"
                  style="height: 1rem !important;"
                >
                  
                </div>
              </div>
            </div>
            <div
              class="col-md-6"
            >
              <div
                class="form-group p-2"
              >
                <label
                  class="text-primary fs-2 fw-800"
                >
                  Fecha revisi贸n
                </label>
                <input
                  class="w100 ng-untouched ng-pristine"
                  disabled=""
                  placeholder=""
                  type="text"
                />
                <div
                  class="text-danger fw-bolder"
                  style="height: 1rem !important;"
                >
                  
                </div>
              </div>
            </div>
            
          </div>
          <div
            class="row p-4"
          >
            <div
              class="col d-flex justify-content-center"
            >
              <app-button>
                <button
                  class="btn btn-primary"
                  disabled=""
                  type="button"
                >
                  Actualizar
                </button>
              </app-button>
            </div>
            
          </div>
        </form>
      </app-form>
    </div>
  </div>
  <app-modal
    class="w25"
  >
    <div
      class="modal-backdrop"
    >
      <div
        class="modal w25"
      >
        
        <div
          class="model-content"
        >
          <div
            class="d-flex justify-content-center mb-2"
          >
            <img
              src="assets/img/happy.svg"
              width="50px"
            />
          </div>
          <p
            class="text-primary fw-800 mb-2"
          >
              
          </p>
          
        </div>
        
      </div>
    </div>
  </app-modal>
</div>
`;
